### ISSUE TITLE
SmolVLA Inference on Jetson Orin (ARM64) Produces Incorrect Results Despite Same Model Working on x86_64

### ISSUE THREAD

**Issue by cjd-hub** [huggingface/lerobot#3636]
### Ticket Type

🐛 Bug Report (Something isn't working)

### Environment & System Info

```Shell
System Info:

Hardware: Jetson Orin (ARM64)
JetPack: 6.2
CUDA: 12.6
PyTorch: 2.5.0 (Jetson native, from jetpack-6.2 channel)
torchvision: 0.20.0
torchcodec: Not available (cannot be installed due to ARM64 / glibc constraints)
LeRobot: 0.4.2
Python environment: venv
Video decoder observed on Jetson: pyav (fallback)

Reference environment (works correctly):
Hardware: x86_64 laptop
PyTorch: 2.7.0
torchvision: 0.21.0
torchcodec: Available and used
Python environment: conda
Video decoder observed: torchcode
```

### Description

**Issue Description:**

I have successfully installed LeRobot 0.4.2 on my Jetson Orin using PyTorch 2.5.0 + torchvision 0.20.0 (the latest versions officially supported on JetPack 6.x). The entire pipeline runs without crashes.

The problem: The model outputs incorrect actions during inference on the real robot. The same SmolVLA model, running with the same scene setup and identical configuration on an x86_64 laptop, successfully completes the task.
**Key observation:**

-     On the Jetson, torchcodec is not available (cannot be installed on ARM64 + JetPack due to glibc / binary compatibility issues). LeRobot automatically falls back to pyav for video decoding.
-     On the x86_64 laptop, torchcodec is installed and used as the video decoding backend.

What works: No runtime errors, model loads successfully, inference loop runs.
What doesn't work: The robot executes seemingly random/unexpected movements instead of the intended task.
**My questions:**

1.Could the difference in video decoding backend (pyav vs torchcodec) be the root cause? 
2.Is torchvision ≥ 0.21 strictly required for correct SmolVLA inference, or is the real requirement that torchcodec must be available? The version constraint might be a proxy for ensuring the right decoding backend.
3.Are there any known differences in tensor output between pyav and torchcodec when decoding the same video frame? For example, differences in:
      (1) Color channel order (RGB vs BGR)
      (2) Pixel value range (0-255 vs 0-1, or different normalization)
      (3) Frame cropping/resizing behavior
      (4)Precision (uint8 vs float32)
4.Is there any recommended way to force torchcodec on Jetson ARM64? Or alternatively, is there a way to make pyav produce bit-identical tensors to torchcodec so that model outputs remain consistent?

**Additional context:**

-       The x86_64 environment uses conda; Jetson uses venv (same Python version 3.10 for both)
-       Both environments use the same model file downloaded from Hugging Face Hub
-       Inference script is identical
-       Physical scene setup (camera positions, object placement) is identical between tests
-       The issue is about inference correctness, not runtime errors or speed

Any insights would be greatly appreciated!

### Relevant logs or stack trace

```Shell

```

### Checklist

- [x] I have searched existing tickets to ensure this isn't a duplicate.
- [ ] I am using the latest version of the `main` branch.
- [ ] I have verified this is not an environment-specific problem.

**Comment by shaoningyu1231** [comment]
Hi @cjd-hub, I ran controlled experiments on the decode-backend hypothesis. I don't think it's your root cause, and there are two other things in your environment I'd rule out first.

- **The decode backend isn't in the live-inference path.** In both v0.4.2 (your version) and v0.6.0 the loop is `robot.get_observation()` → processors → `predict_action`; camera frames go straight from capture into the policy. The torchcodec/pyav backend is only used when reading *dataset* videos (training, replay, offline eval). Unless your script replays dataset frames, the fallback you saw isn't running during inference.
- **On the same machine, the two backends are bit-identical** — 200/200 frames, and also via `decode_video_frames()` across the whole video. No RGB/BGR, value-range, resize or dtype difference to chase.
- **Across architectures there is a real but small difference** (≤3/255, a systematic per-channel shift), and it lives in libswscale's YUV→RGB step, not in the decode: the YUV planes are bit-identical. Propagated through `smolvla_base` in a limited sensitivity probe, it produced only small action changes (~5% of one action std) and did not reproduce grossly wrong behaviour.
- **Note:** your torchvision is below v0.4.2's declared floor, and fp16/bf16 numerics under JetPack's torch 2.5.0.

**Your questions**

1. *Could the decode backend be the root cause?* Very unlikely — mainly for the structural reason above, and in my sensitivity probe the effect was small and did not reproduce grossly wrong behaviour.
2. *Is torchvision ≥0.21 required, or a proxy for torchcodec?* They are independent: torchvision ≥0.21 is an explicit v0.4.2 requirement, while torchcodec was excluded on Linux aarch64 by its dependency marker. So on your Jetson torchcodec was never even attempted — that fallback is by design of the dependency spec, not a glibc failure. Meanwhile your JetPack environment has torchvision 0.20, below the declared floor, which is worth ruling out on its own terms.
3. *Known tensor differences between pyav and torchcodec?* None on the same machine — bit-identical in my runs, including across a cross-FFmpeg-major comparison (4.4 vs 7.1).
4. *Can pyav be made bit-identical to torchcodec, or torchcodec forced on Jetson?* On one machine they already are. For cross-*architecture* parity, the knob is on the x86 side: `av_force_cpu_flags(0)` before the first conversion makes x86 match ARM bit-for-bit, at the cost of the SSSE3 kernel (real conversion-time cost — a parity/debugging knob, not a default). As for forcing torchcodec on Jetson: I would not do it merely for live inference, because that path does not use either video backend. I did not test a manually assembled torchcodec stack on Jetson.

**Evidence, briefly**

Same setup on all hosts: lerobot v0.6.0, `lerobot/svla_so101_pickplace`@`f641879` (AV1 640×480@30fps, md5 verified identical on both machines), av 15.1.0 with libavcodec 61.19.101 / libswscale 8.3.100 everywhere; torchcodec 0.11.1 against system FFmpeg 4.4.2 on x86_64; aarch64 both under QEMU and on real hardware (AWS t4g.small, Graviton2).

The cross-architecture divergence is fully explained by which conversion kernel exists. In FFmpeg n7.1 the AArch64 unscaled YUV→RGB dispatch covers ARGB/RGBA/ABGR/BGRA/GBRP but **not RGB24**, and the generic dispatch tries only PPC/x86/LoongArch initialisers — so YUV420→RGB24 lands on the C function `yuv2rgb_c_24_rgb`. On x86 it selects `yuv420_rgb24_ssse3`. One side runs hand-written SSSE3, the other the C reference: only the C reference exists on AArch64 for this yuv420p→rgb24 conversion in FFmpeg n7.1. Six configurations (x86 and aarch64 × dispatch on/off, plus real hardware) agree with that exactly: **x86 default is the only outlier; everything else is bit-identical to the C reference.**

**What I'd test next on your Jetson** (I don't have Orin hardware, so these are the experiments I'd run to discriminate):

- Force fp32 end-to-end, no autocast/half. If fp32 is correct and fp16/bf16 is wrong, this is torch-2.5.0/JetPack numerics, not decoding.
- Dump a camera frame on both machines under an identical scene and diff the tensors entering the policy. The capture path (e.g. MJPEG decode) has its own architecture-specific SIMD and could show the same signature at larger magnitude.
- With one fixed input tensor saved to disk, compare vision-encoder outputs between Jetson (torch 2.5.0) and x86 — isolates model numerics from the input pipeline.

Happy to share the scripts (staged diff: frames → YUV planes → preprocessed tensors → actions). If useful, this could become a lightweight cross-platform parity check in CI — related to the "systematic policy verification" direction in the roadmap (#3832).

<details>
<summary>Full measurements: cpuflags matrix, divergence statistics, action-level numbers, throughput</summary>

**Backend comparison on x86_64.** Sequential decode of 200 frames: 200/200 identical sha256, max abs diff 0. `decode_video_frames()` on 24 timestamps spanning the video: max diff 0.0.

**Cross-architecture divergence (same av build).** 200/200 RGB frames differ: |diff| ≤ 3, 72.6% of *channel values* off by ±1–2 (that is 99.9% of pixels with at least one channel differing). YUV420p planes: bit-identical, 50/50. x86 host vs x86 container control: 200/200 identical, ruling out distro/glibc effects. Not zero-mean noise but a systematic per-channel shift — signed mean (arm64 − x86) R −0.76, G +0.08, B −1.06 — and stable over time: across all 199 consecutive frame pairs the correlation between successive difference maps
averages 0.83, and where both frames are non-zero at a position, 98.3% of positions keep their sign.

**cpuflags matrix** (`av_force_cpu_flags(0)` applied to the libavutil PyAV has already loaded, before any conversion):

| config | detected → effective flags | RGB output |
|---|---|---|
| x86_64, default dispatch | 1037275 → 1037275 | **differs from every other row** (max 3; R/G/B mean −0.769/−0.074/−1.102) |
| x86_64, cpuflags=0 | 1037275 → 0 | the C reference |
| aarch64 Graviton2, default | 352 → 352 (NEON + ARMV8 + DOTPROD) | bit-identical to the C reference |
| aarch64 Graviton2, cpuflags=0 | 352 → 0 | bit-identical to the C reference |
| aarch64 QEMU, default | 864 → 864 (same + I8MM) | bit-identical to the C reference |
| aarch64 QEMU, cpuflags=0 | 864 → 0 | bit-identical to the C reference |

Emulation was faithful: Graviton2 reproduces the QEMU run bit-for-bit on all 200 frames, and YUV matches x86 50/50 on real silicon. The only difference is cosmetic — QEMU advertised I8MM, which Graviton2 lacks — and it changed nothing.

**Source references** (FFmpeg n7.1):
[aarch64/swscale_unscaled.c](https://ffmpeg.org/doxygen/7.1/aarch64_2swscale__unscaled_8c_source.html)
(no RGB24 among the handled destinations),
[yuv2rgb.c](https://www.ffmpeg.org/doxygen/7.1/yuv2rgb_8c_source.html)
(`ff_yuv2rgb_get_func_ptr` tries PPC/x86/LoongArch only; YUV420→RGB24 →
`yuv2rgb_c_24_rgb`),
[x86/yuv2rgb.c](https://ffmpeg.org/doxygen/7.1/x86_2yuv2rgb_8c_source.html)
(RGB24 → `yuv420_rgb24_ssse3`, asm `ff_yuv_420_rgb24_ssse3`, under
`EXTERNAL_SSSE3`).

**Action-level propagation.** `smolvla_base` (fp32, CPU, same state/task, seeded flow-matching noise, official processor pipeline; same-input determinism check is exact zero), full 50-step chunks at four frames of one episode:

| frame | decode effect (max over chunk) | one-frame scene advance | ratio |
|---|---|---|---|
| 10 | 0.0496 std | 0.0054 std | 9.2× |
| 30 | 0.0401 std | 0.0077 std | 5.2× |
| 60 | 0.0157 std | 0.0141 std | 1.1× |
| 100 | 0.0550 std | 0.1126 std | 0.5× |

Calibration from ground-truth actions in the same episode (median of per-step max, action-std units): 0.026 over 1 frame, 0.141 over 5, 0.521 over 15, 2.12 over 50 — one chunk horizon. Precision is a much smaller effect: fp32 vs bf16-autocast on an identical frame differs by 0.0004 std, and the decode-induced difference is essentially unchanged under bf16 (0.0160 vs 0.0162 std), so reduced precision did not amplify input sensitivity here. Caveats: one policy, one episode, four frames, fixed seed, CPU fp32, and the same camera image copied into
all three of SmolVLA's image slots — a sensitivity probe, not a rollout.

**Throughput of the parity knob** (cross-validation of the source reading, not the argument). Disabling dispatch on real ARM changes nothing measurable (+0.2% end-to-end, +0.9% on the conversion step — there is no accelerated path to disable). On x86 it clearly costs: conversion step +60% and +253% across two rounds; comparing the per-configuration medians gives +145%. End-to-end, ~20%. Direction is reliable, the exact ratio is not — the ARM host is a burstable t4g instance, and my auto-threaded x86 runs were unusable (decode-only noise floor ~10%). A defensible figure would need a dedicated harness.

</details>

**Comment by ravediamond** [comment]
@shaoningyu1231 finally have real Orin hardware — ran your first two suggested experiments (fp32-only, same-machine decode diff), happy to share numbers once I've sanity-checked them.

For the third one (fixed-input vision-encoder comparison against x86 torch 2.7.0) I don't have x86+CUDA hardware to generate a matched reference — would you be able to share the saved input tensor + your vision-encoder output, or the script itself? Would let me run the same comparison on Orin without introducing a new methodology mismatch.
